src/request/plan.rs

```diff
@@ -315,102 +301,64 @@ where
     }
 }
 
-/// When executed, the plan clones the keys and execute its inner plan, then
-/// returns `(keys, response)`.
+/// When executed, the plan clones the shard and execute its inner plan, then
+/// returns `(shard, response)`.
 ///
-/// It's useful when the information of keys are lost in the response but needed
+/// It's useful when the information of shard are lost in the response but needed
 /// for processing.
-pub struct PreserveKey<P: Plan + HasKeys> {
+pub struct PreserveShard<P: Plan + Shardable> {
     pub inner: P,
 }
 
-impl<P: Plan + HasKeys> Clone for PreserveKey<P> {
+impl<P: Plan + Shardable> Clone for PreserveShard<P> {
     fn clone(&self) -> Self {
-        PreserveKey {
+        PreserveShard {
             inner: self.inner.clone(),
         }
     }
 }
 
 #[async_trait]
-impl<P> Plan for PreserveKey<P>
+impl<P> Plan for PreserveShard<P>
 where
-    P: Plan + HasKeys,
+    P: Plan + Shardable,
 {
-    type Result = ResponseAndKeys<P::Result>;
+    type Result = ResponseWithShard<P::Result, P::Shard>;
 
     async fn execute(&self) -> Result<Self::Result> {
-        let keys = self.inner.get_keys();
+        let shard = self.inner.get_shard();
         let res = self.inner.execute().await?;
-        Ok(ResponseAndKeys(res, keys))
+        Ok(ResponseWithShard(res, shard))
     }
 }
 
-pub trait HasKeys {
-    fn get_keys(&self) -> Vec<Key>;
-}
-
 // contains a response and the corresponding keys
 // currently only used for matching keys and values in pessimistic lock requests
```

Comment:
The comment needs update. And I think we can add a hint saying that the order of the elements in the response and in the shard must match.